### PR TITLE
dashboard intro tutorial

### PR DESCRIPTION
**Pull request by tgberkeley**

### CODE REVIEW

File: docs/getting-started/dashboard_tutorial.md [hunk 1]
```diff
@@ -0,0 +1,1788 @@
...
+
+During this tutorial you will build a small data dashboard, where you can input data and it will be rendered in table and a graph. This tutorial does not assume any existing Reflex knowledge, but we do recommend checking out the quick [Basics Guide]({docs.getting_started.basics.path}) first. 
+
+The techniques you’ll learn in the tutorial are fundamental to building any Reflex app, and fully understanding it will give you a deep understanding of Reflex.
+
+
+This tutorial is divided into several sections:
+
+- Setup for the tutorial will give you a starting point to follow the tutorial.
+- Overview will teach you the fundamentals of Reflex: components and props.
```
Comment (amsraman) [inline docs/getting-started/dashboard_tutorial.md:16]: ehh personal preference but I would maybe format these summary bullet points a little differently. something like

- **Setup for the Tutorial**: A starting point to follow the tutorial
- **Overview**: The fundamentals of Reflex UI (components and props)
- **Showing Dynamic Data**: How to use state to render dynamic data. 

I dunno, something that clearly shows that these are subsection headings that you're mentioning, and separating out the description maybe.

File: docs/getting-started/dashboard_tutorial.md [hunk 2]
```diff
@@ -0,0 +1,1788 @@
...
+- Final Cleanup and Conclusion will explain how to further customize your app and add some extra styling to it.
+
+### What are you building?
+
+In this tutorial, you are building an interactive data dashboard with Reflex.
+
+You can see what the finished app and code will look like here:
+
+
+```python exec
```
Comment (amsraman) [inline docs/getting-started/dashboard_tutorial.md:29]: Maybe would be nice if the dashboard can be surrounded by some box or border. Just to clearly separate it out from the text in the page.

File: docs/getting-started/dashboard_tutorial.md [hunk 3]
```diff
@@ -0,0 +1,1788 @@
...
+        ),
+        graph5(),
+        align="center",
+        width="100%",
+        on_mouse_enter=State5.transform_data
+    )
+```
+
+```python
+import reflex as rx
```
Comment (amsraman) [inline docs/getting-started/dashboard_tutorial.md:178]: Maybe would be less cluttered to put the code inside a drop down? Or I guess maybe your big point is that the code is so short that you wouldn't need to do that.

File: docs/getting-started/dashboard_tutorial.md [hunk 4]
```diff
@@ -0,0 +1,1788 @@
...
+    on_load=State.transform_data,
+)
+```
+
+Don't worry if you don't understand the code above, in this tutorial we are going to walk you through the whole thing step by step.
+
+
+## Setup for the tutorial
+
+Check out the [installation docs]({docs.getting_started.installation.path}) to get Reflex set up on your machine. Follow these to create a folder called `dashboard_tutorial`, which you will `cd` into and then `pip install reflex`.
```
Comment (amsraman) [inline docs/getting-started/dashboard_tutorial.md:339]: I guess you don't need to cd into the folder to pip install reflex

File: docs/getting-started/dashboard_tutorial.md [hunk 5]
```diff
@@ -0,0 +1,1788 @@
...
+We will choose template `0` when we run `reflex init` to get the blank template. Finally run `reflex run` to start the app and confirm everything is set up correctly.
+
+
+## Overview
+
+Now that you’re set up, let’s get an overview of Reflex!
+
+### Inspecting the starter code
+
+Within our `dashboard_tutorial` folder we just `cd`'d into, there is a `rxconfig.py` file that contains the configuration for our Reflex app. ([config docs]({docs.getting_started.configuration.path}) for more information)
```
Comment (amsraman) [inline docs/getting-started/dashboard_tutorial.md:350]: This renders as a stray space appearing between the paren and ``config docs''.

File: docs/getting-started/dashboard_tutorial.md [hunk 6]
```diff
@@ -0,0 +1,1788 @@
...
+
+There is also an `assets` folder where static files such as images and stylesheets can be placed to be referenced within your app. ([asset docs]({docs.assets.referencing_assets.path}) for more information)
```
Comment (amsraman) [inline docs/getting-started/dashboard_tutorial.md:352]: A stray space renders before ``asset docs''

File: docs/getting-started/dashboard_tutorial.md [hunk 7]
```diff
@@ -0,0 +1,1788 @@
...
+
+Most importantly there is a folder also called `dashboard_tutorial` which contains all the code for your app. Inside of this folder there is a file named `dashboard_tutorial.py`. To begin this tutorial we will delete all the code in this file so that we can start from scratch and explain every step as we go.
+
+The first thing we need to do is import `reflex`. Once we have done this we can create a component, which is a a piece of reusable code that represents a part of a user interface. Components are used to render, manage, and update the UI elements in your application.
```
Comment (amsraman) [inline docs/getting-started/dashboard_tutorial.md:356]: a a piece of reusable code. also maybe rephrase this as "a reusable piece of UI code"? idk, your choice

File: docs/getting-started/dashboard_tutorial.md [hunk 8]
```diff
@@ -0,0 +1,1788 @@
...
+
+Within our `dashboard_tutorial` folder we just `cd`'d into, there is a `rxconfig.py` file that contains the configuration for our Reflex app. ([config docs]({docs.getting_started.configuration.path}) for more information)
+
+There is also an `assets` folder where static files such as images and stylesheets can be placed to be referenced within your app. ([asset docs]({docs.assets.referencing_assets.path}) for more information)
+
+Most importantly there is a folder also called `dashboard_tutorial` which contains all the code for your app. Inside of this folder there is a file named `dashboard_tutorial.py`. To begin this tutorial we will delete all the code in this file so that we can start from scratch and explain every step as we go.
+
+The first thing we need to do is import `reflex`. Once we have done this we can create a component, which is a a piece of reusable code that represents a part of a user interface. Components are used to render, manage, and update the UI elements in your application. 
+
+Let's look at the example below. Here we have a function called `index` that returns a `text` component (an in-built Reflex UI component) that displays the text "Hello World!".
```
Comment (amsraman) [inline docs/getting-started/dashboard_tutorial.md:358]: Maybe mention that the main function *has* to be called index? Otherwise they might call it something else and see an error they don't understand.

File: docs/getting-started/dashboard_tutorial.md [hunk 9]
```diff
@@ -0,0 +1,1788 @@
...
+                rx.table.cell("zahra@example.com"),
+                rx.table.cell("Female"),
+            ),
+        ),
+    )
+```
+
+Components in Reflex have `props`, which can be used to customize the component and are passed in as keyword arguments to the component function. 
+
+The `rx.table.root` component has for example the `variant` and `size` props, which customize the table as seen below.
```
Comment (amsraman) [inline docs/getting-started/dashboard_tutorial.md:438]: Size makes sense, maybe as a reader I don't understand what variant is or what values it can be. I don't know though, maybe I naturally try to understand too much on the first read.

File: docs/getting-started/dashboard_tutorial.md [hunk 10]
```diff
@@ -0,0 +1,1788 @@
...
+
+This is where `State` comes in. `State` is a Python class that stores variables that can change when the app is running, as well as the functions that can change those variables.
+
+To define a state class, subclass `rx.State` and define fields that store the state of your app. The state variables (vars) should have a type annotation, and can be initialized with a default value. Check out the [basics]({docs.getting_started.basics.path}) section for a simple example of how state works.
+
+
+In the example below we define a `State` class called `State` that has a variable called `users` that is a list of lists. Each list in the `users` list represents a user and contains their name, email and gender.
+
+```python
+class State(rx.State):
```
Comment (amsraman) [inline docs/getting-started/dashboard_tutorial.md:505]: I think my general opinion is that introducing state classes with a class called ``State'' is weird. Because then a question arises as to whether that keyword should especially used for naming the State class or not. But I don't know, if this is the standard for the rest of the docs then it's fine.

File: docs/getting-started/dashboard_tutorial.md [hunk 11]
```diff
@@ -0,0 +1,1788 @@
...
+    ]
+```
+
+To iterate over a state var that is a list, we use the [`rx.foreach`]({docs.components.rendering_iterables.path}) function to render a list of components. The `rx.foreach` component takes an `iterable` (list, tuple or dict) and a `function` that renders each item in the `iterable`.
+
+Here the render function is `show_user` which takes in a single user and returns a `table.row` component that displays the users name, email and gender.
+
+```python exec
+class State1(rx.State):
+    users: list[list] = [
```
Comment (amsraman) [inline docs/getting-started/dashboard_tutorial.md:518]: Why not list[list[str]]? Eh.

File: docs/getting-started/dashboard_tutorial.md [hunk 12]
```diff
@@ -0,0 +1,1788 @@
...
+
+```python
+class State(rx.State):
+    users: list[list] = [
+        ["Danilo Sousa", "danilo@example.com", "Male"],
+        ["Zahra Ambessa", "zahra@example.com", "Female"],
+    ]
+```
+
+To iterate over a state var that is a list, we use the [`rx.foreach`]({docs.components.rendering_iterables.path}) function to render a list of components. The `rx.foreach` component takes an `iterable` (list, tuple or dict) and a `function` that renders each item in the `iterable`.
```
Comment (amsraman) [inline docs/getting-started/dashboard_tutorial.md:512]: As a new user, I wonder why I can't just splat a for loop here. Maybe a quick mention/redirect to that section of the basic docs?

File: docs/getting-started/dashboard_tutorial.md [hunk 13]
```diff
@@ -0,0 +1,1788 @@
...
+
+As you can see the output above looks the same as before, except now the data is no longer static and can change with user input to the app.
+
+### Using a proper class structure for our data
+
+So far our data has been defined in a list of lists, where the data is accessed by index i.e. `user[0]`, `user[1]`. This is not very maintainable as our app gets bigger. 
+
+A better way to structure our data in Reflex is to use a class to represent a user. This way we can access the data using attributes i.e. `user.name`, `user.email`.
+
+In Reflex when we create these classes to showcase our data, the class must inherit from `rx.Base`.
```
Comment (amsraman) [inline docs/getting-started/dashboard_tutorial.md:594]: Why should I inherit from rx.Base? What usage of this class would necessitate that. Because there doesn't seem anything special about it right now

File: docs/getting-started/dashboard_tutorial.md [hunk 14]
```diff
@@ -0,0 +1,1788 @@
...
+        ),
+    ),
+)
+```
+
+Now you have probably realised that we have all the form fields, but we have no way to submit the form. We can add a submit button to the form by adding a `rx.button` component to the `vstack` component. The `rx.button` component takes in the text that is displayed on the button and the `type` prop which is the type of button. The `type` prop is set to `submit` so that the form is submitted when the button is clicked. 
+
+In addition to this we need a way to update the `users` state variable when the form is submitted. All state changes are handled through functions in the state class, called [event handlers]({docs.events.events_overview.path}).
+
+Components have special props called event triggers, such as `on_submit`, that can be used to make components interactive. Event triggers connect components to event handlers, which update the state.
```
Comment (amsraman) [inline docs/getting-started/dashboard_tutorial.md:739]: The way this is described for me is... eh. What does on_submit take in a dictionary? Are there events? What do they take in? Again, maybe I ask too many questions as a new user haha

File: docs/getting-started/dashboard_tutorial.md [hunk 15]
```diff
@@ -0,0 +1,1788 @@
...
+            rx.button("Submit", type="submit"),
+        ),
+        on_submit=State.add_user,
+        reset_on_submit=True,
+    )
+```
+
+Finally we must add the new `form()` component we have defined to the `index()` function so that the form is rendered on the page.
+
+Below is the full code for the app so far. If you try this form out you will see that you can add new users to the table by filling out the form and clicking the submit button. The form data will also appear as a toast on the screen when submitted.
```
Comment (amsraman) [inline docs/getting-started/dashboard_tutorial.md:777]: What is a toast? Maybe say it will appear "as a small window in the corner" or something.

File: docs/getting-started/dashboard_tutorial.md [hunk 16]
```diff
@@ -0,0 +1,1788 @@
...
+            variant="surface",
+            size="3",
+        ),
+    )
+```
+
+
+### Putting the Form in an Overlay
+
+In Reflex we like to make the user interaction as intuitive as possible. Placing the form we just constructed in an overlay creates a focused interaction by dimming the background, and ensures a cleaner layout when you have multiple action points such as editing and deleting as well.
```
Comment (amsraman) [inline docs/getting-started/dashboard_tutorial.md:913]: What.... is an overlay? Why do I want this? You kind of describe it, but ehh (maybe I have low attention span)

I think in general, maybe show the product of this component first. Then as user I'm like "hey, that's cool! how can I add that?". Then say that it's an overlay and show how to use it.

Otherwise it's like "huh. why do I need this? what is this?". By this point I might tune out. If I don't, by the end "ohhh that's what it does. ok sure that's cool".

Also add a comma after "In Reflex".

File: docs/getting-started/dashboard_tutorial.md [hunk 17]
```diff
@@ -0,0 +1,1788 @@
...
+```python
+rx.dialog.trigger(
+    rx.button(
+        rx.icon("plus", size=26),
+        rx.text("Add User", size="4"),
+    ),
+)
+```
+
+After the trigger we have the `rx.dialog.content` which contains everything within our dialog, including a title, a description and our form. The key part of the form within the dialog is that it now has two `rx.dialog.close` components. The first is to just close the dialog without submitting the form and he second is to close the dialog and submit the form as shown below.
```
Comment (amsraman) [inline docs/getting-started/dashboard_tutorial.md:926]: Two close components? Why? Ahhh I'm confused now.

Again, maybe first come from the behavior we want (we want two different ways to close it), and then say we should have two close components as a result (instead of the other way around).

Again, if as a user I saw the final product first, it would more naturally make sense.

File: docs/getting-started/dashboard_tutorial.md [hunk 18]
```diff
@@ -0,0 +1,1788 @@
...
+```
+
+
+## Plotting Data in a Graph
+
+The last part of this tutorial is to plot the user data in a graph. We will use Reflex's built-in graphing library recharts to plot the number of users of each gender. 
+
+### Transforming the data for the graph
+
+The graphing components in Reflex expect to take in a list of dictionaries. Each dictionary represents a data point on the graph and contains the x and y values. We will create a new event handler in the state called `transform_data` to transform the user data into the format that the graphing components expect. We must also create a new state variable called `users_for_graph` to store the transformed data, which will be used to render the graph.
```
Comment (amsraman) [inline docs/getting-started/dashboard_tutorial.md:1197]: This is slightly confusing. So, is transform_data automatically run? I don't see you calling anywhere. Is it some kind of special thing that's particular to the graph component?

If so, mention that. And also how exactly it hooks up to it.

File: docs/getting-started/dashboard_tutorial.md [hunk 19]
```diff
@@ -0,0 +1,1788 @@
...
+        ]
+```
+
+As we can see above the `transform_data` event handler uses the `Counter` class from the `collections` module to count the number of users of each gender. We then create a list of dictionaries from this which we set to the state var `users_for_graph`. 
+
+Finally we can see that whenever we add a new user through submitting the form, we call the `transform_data` event handler to update the `users_for_graph` state variable.
+
+### Rendering the graph
+
+We use the `rx.recharts.bar_chart` component to render the graph. We pass through the state variable for our graphing data as `data=State.users_for_graph`. We alos pass in a `rx.recharts.bar` component which represents the bars on the graph. The `rx.recharts.bar` component takes in the `data_key` prop which is the key in the data dictionary that represents the y value of the bar. The `stroke` and `fill` props are used to set the color of the bars.
```
Comment (amsraman) [inline docs/getting-started/dashboard_tutorial.md:1232]: "alos pass"

File: docs/getting-started/dashboard_tutorial.md [hunk 20]
```diff
@@ -0,0 +1,1788 @@
...
+
+```python
+app = rx.App(
+    theme=rx.theme(
+        radius="full", accent_color="grass"
+    ),
+)
+```
+
+Unfortunately in this tutoial here we cannot actually apply this to the live example on the page, but if you copy and paste the code below into a reflex app locally you can see if in action.
```
Comment (amsraman) [inline docs/getting-started/dashboard_tutorial.md:1576]: tutoial

File: docs/getting-started/dashboard_tutorial.md [hunk 21]
```diff
@@ -0,0 +1,1788 @@
...
+
+### Revisiting app=rx.App()
+
+At the beginning of the tutorial we also mentioned that we defined our app using `app=rx.App()`. We can also pass in some props to the `rx.App` component to customize the app.
+
+The most important one is `theme` which allows you to customize the look and feel of the app. The `theme` prop takes in an `rx.theme` component which has several props that can be set. 
+
+The `radius` prop sets the global radius value for the app that is inherited by all components that have a `radius` prop. It can be overwritten locally for a specific component by manually setting the `radius` prop.
+
+The `accent_color` prop sets the accent color of the app. Check out other options for the accent color [here]({docs.library.other.theme.path}).
```
Comment (amsraman) [inline docs/getting-started/dashboard_tutorial.md:1565]: These are a few of the props; where can I see the rest? And what else can I set these props to? Maybe a redirect to some page idk.

File: docs/getting-started/dashboard_tutorial.md [hunk 22]
```diff
@@ -0,0 +1,1788 @@
...
+
+In addition to the above we have we have 
+
+- explored state to allow you to show dynamic data that changes over time
+- explored events to allow you to make your app interactive and respond to user actions
+- added styling to the app to make it look better
+
+
+
+## Advanced Section (Hooking this up to a Database)
```
Comment (amsraman) [inline docs/getting-started/dashboard_tutorial.md:1788]: This would be quite nice, and can probably be explained concisely enough.

**Review by amsraman — COMMENTED** [review]
Here are my comments. Some might be a bit nitpicky haha

File: docs/getting-started/dashboard_tutorial.md (hunk 1 above)
Comment (tgberkeley) [reply docs/getting-started/dashboard_tutorial.md:16]: fixed

File: docs/getting-started/dashboard_tutorial.md (hunk 3 above)
Comment (tgberkeley) [reply docs/getting-started/dashboard_tutorial.md:178]: waiting for another PR in reflex web for this to be possible, let's maybe do it as a follow up

File: docs/getting-started/dashboard_tutorial.md (hunk 2 above)
Comment (tgberkeley) [reply docs/getting-started/dashboard_tutorial.md:29]: fixed for all that needed it

File: docs/getting-started/dashboard_tutorial.md (hunk 4 above)
Comment (tgberkeley) [reply docs/getting-started/dashboard_tutorial.md:339]: I changed it to remove the word then as you are right the order does not matter

File: docs/getting-started/dashboard_tutorial.md (hunk 7 above)
Comment (tgberkeley) [reply docs/getting-started/dashboard_tutorial.md:356]: changed it to: which is a reusable piece of user interface code

File: docs/getting-started/dashboard_tutorial.md (hunk 5 above)
Comment (tgberkeley) [reply docs/getting-started/dashboard_tutorial.md:350]: something to explore in another PR as is an issue with the .path

File: docs/getting-started/dashboard_tutorial.md (hunk 6 above)
Comment (tgberkeley) [reply docs/getting-started/dashboard_tutorial.md:352]: same as above

File: docs/getting-started/dashboard_tutorial.md (hunk 8 above)
Comment (tgberkeley) [reply docs/getting-started/dashboard_tutorial.md:358]: added this line: The function name (in this example `index`) which defines the component, must be what we pass into the `add_page`.

File: docs/getting-started/dashboard_tutorial.md (hunk 9 above)
Comment (tgberkeley) [reply docs/getting-started/dashboard_tutorial.md:438]: I think it is okay, it is something they will need to learn so I think it gives a nice simple intro to it

File: docs/getting-started/dashboard_tutorial.md (hunk 10 above)
Comment (tgberkeley) [reply docs/getting-started/dashboard_tutorial.md:505]: I think this is fairly standard as it's what we do in the introduction and basics

File: docs/getting-started/dashboard_tutorial.md (hunk 11 above)
Comment (tgberkeley) [reply docs/getting-started/dashboard_tutorial.md:518]: very true, will change now

File: docs/getting-started/dashboard_tutorial.md (hunk 12 above)
Comment (tgberkeley) [reply docs/getting-started/dashboard_tutorial.md:512]: this is a fantastic point, I have added a link to the runtime vs compile time part of the basics section

File: docs/getting-started/dashboard_tutorial.md (hunk 13 above)
Comment (tgberkeley) [reply docs/getting-started/dashboard_tutorial.md:594]: added more information here (think we should add a follow up with a link to a doc with more rx.base)

File: docs/getting-started/dashboard_tutorial.md (hunk 15 above)
Comment (tgberkeley) [reply docs/getting-started/dashboard_tutorial.md:777]: added

File: docs/getting-started/dashboard_tutorial.md (hunk 14 above)
Comment (tgberkeley) [reply docs/getting-started/dashboard_tutorial.md:739]: fixed

File: docs/getting-started/dashboard_tutorial.md (hunk 16 above)
Comment (tgberkeley) [reply docs/getting-started/dashboard_tutorial.md:913]: I think we see the use of the overlay in the original demo at the top of the page

File: docs/getting-started/dashboard_tutorial.md (hunk 17 above)
Comment (tgberkeley) [reply docs/getting-started/dashboard_tutorial.md:926]: fixed

File: docs/getting-started/dashboard_tutorial.md (hunk 18 above)
Comment (tgberkeley) [reply docs/getting-started/dashboard_tutorial.md:1197]: added the info to the section after this

File: docs/getting-started/dashboard_tutorial.md (hunk 19 above)
Comment (tgberkeley) [reply docs/getting-started/dashboard_tutorial.md:1232]: fixed

File: docs/getting-started/dashboard_tutorial.md (hunk 20 above)
Comment (tgberkeley) [reply docs/getting-started/dashboard_tutorial.md:1576]: fixed

File: docs/getting-started/dashboard_tutorial.md (hunk 21 above)
Comment (tgberkeley) [reply docs/getting-started/dashboard_tutorial.md:1565]: added link to theming docs

File: docs/getting-started/dashboard_tutorial.md (hunk 22 above)
Comment (tgberkeley) [reply docs/getting-started/dashboard_tutorial.md:1788]: will add in follow up

File: docs/getting-started/dashboard_tutorial.md (hunk 8 above)
Comment (amsraman) [reply docs/getting-started/dashboard_tutorial.md:361]: Nice! On a second readthrough, maybe bolding the word component on its first mention would be nice.

File: docs/getting-started/dashboard_tutorial.md (hunk 9 above)
Comment (amsraman) [reply docs/getting-started/dashboard_tutorial.md:448]: Sure

File: docs/getting-started/dashboard_tutorial.md (hunk 12 above)
Comment (amsraman) [reply docs/getting-started/dashboard_tutorial.md:525]: Nice!

File: docs/getting-started/dashboard_tutorial.md (hunk 13 above)
Comment (amsraman) [reply docs/getting-started/dashboard_tutorial.md:615]: Hmm maybe switch the ``show_user'' paragraph with the one after it. Or just merge this new paragraph with the one that introduces rx.base, for concision.

File: docs/getting-started/dashboard_tutorial.md (hunk 18 above)
Comment (amsraman) [reply docs/getting-started/dashboard_tutorial.md:1229]: yeah nice! I also think I just missed this function being called on my first readthrough. oops

**Review by amsraman — COMMENTED** [review]
Nice, it looks much better!

One other thing that might be nice: you sort of progressively show the code as this tutorial goes on: it would be nice if there was some way to highlight the parts of it that are new or were just added. Not very important though

**Review by amsraman — APPROVED** [review]
